### PR TITLE
a-o-i: Add support for Atomic Registry Installs

### PR DESCRIPTION
**Pull request by smunilla**
Add the Registry deployment type as an option in the quick installer.

Supersedes #2318

### CODE REVIEW

File: roles/cockpit-ui/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+---
+- name: Install cockpit-ui
+  action: "{{ ansible_pkg_mgr }} name={{ item }} state=present"
+  when: not openshift.common.is_containerized | bool
```
Comment (sdodson) [inline roles/cockpit-ui/tasks/main.yml]: Needs either with_items: list or just change {{ item }} to the name of the package. Also, maybe we can use the existing cockpit role?

**Comment by smunilla** [comment]
So this should work or at least do something. You can either:
1) run the quick installer interactively and select Registry as the deployment type
2) run the quick installer unattended with 'variant_subtype: regisitry' in the installer config
3) run ansible directly with 'deployment_subtype=registry' in the inventory.

cc/ @brenton

**Comment by brenton** [comment]
Awesome.  I'll try this out now.

**Comment by aweiteka** [comment]
cc @brianwcook for an openshift-based registry install. This is probably the best choice for multi-host HA.

**Comment by Jmainguy** [comment]
Used quick installer, got this error.

https://paste.fedoraproject.org/404586/

**Comment by abutcher** [comment]
@Jmainguy did you have this commit in your tree? ca0abdd8ebc5b0dfe2ce0c5775f09599bf830934

**Comment by smunilla** [comment]
@Jmainguy any chance you still have /tmp/ansible.log from that run?

File: utils/src/ooinstall/cli_installer.py [hunk 2]
```diff
@@ -742,6 +744,15 @@ def get_hosts_to_run_on(oo_cfg, callback_facts, unattended, force, verbose):
 
     return hosts_to_run_on, callback_facts
 
+def set_infra_nodes(hosts):
+    if all(host.is_master() for host in hosts):
+        infra_list = hosts
+    else:
+        infra_list = hosts[:2]
```
Comment (abutcher) [inline utils/src/ooinstall/cli_installer.py]: Will this potentially select unscheduleable masters?

**Comment by aweiteka** [comment]
@smunilla where is the [cockpit template](https://github.com/cockpit-project/cockpit/blob/master/containers/openshift-cockpit.template) coming from? I assumed it would land in our examples dir (roles/openshift_examples/files/examples). Since this is a cluster admin template, it feels like it should be installed under "default" project. @brenton thoughts?

**Comment by brenton** [comment]
Yeah, I think it makes the most sense to add it to the openshift_examples role where the rest of the templates are.

**Comment by smunilla** [comment]
@aweiteka Just did the needful regarding the template. It probably needs some tweaking, but an early opinion would be good.

**Comment by smunilla** [comment]
@aweiteka We haven't had the best luck running the template either manually or through this code. Some more guidance would be appreciated.

File: utils/src/ooinstall/cli_installer.py (hunk 2 above)
Comment (smunilla) [inline utils/src/ooinstall/cli_installer.py]: You're right. i refactored my own code and left out the nodes-only part. Fixing.

File: roles/openshift_examples/files/examples/v1.3/cockpit-templates/enterprise/openshift-cockpit.template [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+                 "protocol": "TCP",
+                 "port": 80,
+                 "targetPort": 9090
+             },
+             {
+                 "name": "https",
+                 "protocol": "TCP",
+                 "port": 443,
+                 "targetPort": 9090
+             }],
```
Comment (detiber) [inline roles/openshift_examples/files/examples/v1.3/cockpit-templates/enterprise/openshift-cockpit.template]: Does port 9000 conflict with the locally running cockpit instance? 

Should both http and https both be pointing to targetPort 9090?

File: playbooks/common/openshift-cluster/additional_config.yml [hunk 4]
```diff
@@ -23,12 +23,8 @@
...
   - role: cockpit
     when: not openshift.common.is_atomic and ( deployment_type in ['atomic-enterprise','openshift-enterprise'] ) and
-      (osm_use_cockpit | bool or osm_use_cockpit is undefined )
-  - role: flannel_register
-    when: openshift.common.use_flannel | bool
-
-- name: Cockpit UI configuration
-  hosts: oo_first_master
-  roles:
+      (osm_use_cockpit | bool or osm_use_cockpit is undefined ) and ( deployment_subtype != 'registry' )
```
Comment (detiber) [inline playbooks/common/openshift-cluster/additional_config.yml]: Condition probably should be:
```
osm_use_cockpit | default(false) | bool and deployment_subtype != 'registry'
```
The deployment_subtype could be moved to the default filter, but I think that would be a bit counter-intuitive to read.

That said, I think this is a bit confusing to follow that the setting of deployment_subtype would override the osm_use_cockpit variable in that way.

Would it make sense to possibly deploy the cockpit container by default on all systems as opposed to doing the package based deployment?

File: roles/cockpit-ui/tasks/main.yml [hunk 5]
```diff
@@ -1,12 +1,9 @@
...
-  with_items:
-    - cockpit-ui
-  when: not openshift.common.is_containerized | bool
-
 - name: Cockpit template
-  command: oc new-app {{ cockpit_ui_base }} -p OPENSHIFT_OAUTH_PROVIDER_URL={{ openshift_https_proxy }} -p COCKPIT_KUBE_URL={{ openshift_https_proxy }} -p KUBERNETES_SERVICE_HOST={{ openshift_https_proxy }} -p KUBERNETES_SERVICE_PORT=8443 -p REGISTRY_HOST=REGISTRYHOSTNAME:5000 -p OPENSHIFT_OAUTH_CLIENT_ID=cockpit-oauth-client -p KUBERNETES_INSECURE=true 
+  command: oc new-app {{ cockpit_ui_base }} -p OPENSHIFT_OAUTH_PROVIDER_URL={{ openshift_https_proxy }} -p COCKPIT_KUBE_URL={{ openshift_https_proxy }} -p KUBERNETES_SERVICE_HOST={{ openshift_https_proxy }} -p KUBERNETES_SERVICE_PORT=8443 -p REGISTRY_HOST=REGISTRYHOSTNAME:5000 -p OPENSHIFT_OAUTH_CLIENT_ID=cockpit-oauth-client -p KUBERNETES_INSECURE=true
+
+- name: Patch Cockpit routes
+  command:
```
Comment (detiber) [inline roles/cockpit-ui/tasks/main.yml]: This task seems to be missing arguments.

File: roles/cockpit-ui/tasks/main.yml [hunk 6]
```diff
@@ -0,0 +1,22 @@
+---
```
Comment (smunilla) [inline roles/cockpit-ui/tasks/main.yml]: @aweiteka Does this look like the right thing to do here?

File: roles/cockpit-ui/tasks/main.yml [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+- name: Expose registry with route
+  command: oc expose service docker-registry
+
+- name: Install Cockpit template
+  command: oc create -f registry-console.yaml -n default
```
Comment (aweiteka) [inline roles/cockpit-ui/tasks/main.yml]: This file is a reference to one of the templates in #2318

File: roles/cockpit-ui/tasks/main.yml [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+- name: Create passthrough route for Registry
+  command: oc create route passthrough --service registry-console --port registry-console -n default
+
+- name: Deploy Registry
+  command: oc new-app -n default --template=registry-console -p OPENSHIFT_OAUTH_PROVIDER_URL="{{ openshift_https_proxy }}:8443",REGISTRY_HOST=$(oc get route docker-registry -n default --template='{{ .spec.host }}'),COCKPIT_KUBE_URL=$(oc get route registry-console -n default --template='https://{{ .spec.host }}')
+
+- name: Patch Cockpit routes
```
Comment (aweiteka) [inline roles/cockpit-ui/tasks/main.yml]: Not necessary. We're using a passthrough route with auto-generated SSL certs. Custom certs are post install per docs PR https://github.com/openshift/openshift-docs/pull/2696

**Comment by aweiteka** [comment]
Templates should be removed from this PR. Use templates from #2318

File: roles/cockpit-ui/tasks/main.yml (hunk 7 above)
Comment (aweiteka) [inline roles/cockpit-ui/tasks/main.yml]: Remove this step if example templates are already installed via previous playbook.

File: roles/openshift_examples/files/examples/v1.3/infrastructure-templates/enterprise/registry-console.yaml [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+    apiVersion: v1
+    metadata:
+      name: "${OPENSHIFT_OAUTH_CLIENT_ID}"
+      respondWithChallenges: false
+    secret: "${OPENSHIFT_OAUTH_CLIENT_SECRET}"
+    redirectURIs:
+      - "${COCKPIT_KUBE_URL}"
+parameters:
+  - description: 'Specify "registry/repository" prefix for container image; e.g. for "registry.access.redhat.com/openshift3/registry-console:latest", set prefix "registry.access.redhat.com/openshift3/"'
+    name: IMAGE_PREFIX
```
Comment (aweiteka) [inline roles/openshift_examples/files/examples/v1.3/infrastructure-templates/enterprise/registry-console.yaml]: @smunilla are these params exposed in this playbook? Is there a way to pass in override values for testing?

Comment (smunilla) [inline roles/openshift_examples/files/examples/v1.3/infrastructure-templates/enterprise/registry-console.yaml]: @aweiteka Yeah, these are exposed, I'll work up some examples on how.

**Comment by smunilla** [comment]
@aweiteka @brenton: With a lot of help from @abutcher, we got this deploying the cockpit-ui late yesterday. More eye-on testing would be good at this point :)

**Comment by brenton** [comment]
@smunilla, what's the latest status on this PR?

**Comment by smunilla** [comment]
@brenton I sent a build off to QE last night. They've provided some feedback on the card.

**Comment by smunilla** [comment]
@brenton And for a double update. We've addressed QE's concerns and should be ready to merge.

**Comment by smunilla** [comment]
Supersedes #2318

**Comment by brenton** [comment]
Anything blocking the merge now?

**Comment by detiber** [comment]
aos-ci-tests

**Comment by smunilla** [comment]
Based on feedback from QE, I think we're good to go here.

**Comment by smunilla** [comment]
aos-ci-test

**Comment by smunilla** [comment]
aos-ci-test

**Comment by smunilla** [comment]
aos-ci-test

**Pull request merged**